service/src/main/kotlin/xyz/lebkuchenfm/external/youtube/YoutubeClient.kt

```diff
@@ -1,44 +1,80 @@
 package xyz.lebkuchenfm.external.youtube
 
+import com.github.michaelbull.result.Err
+import com.github.michaelbull.result.Ok
+import com.github.michaelbull.result.Result
+import io.github.oshai.kotlinlogging.KotlinLogging
 import io.ktor.client.HttpClient
 import io.ktor.client.call.body
 import io.ktor.client.engine.okhttp.OkHttp
+import io.ktor.client.plugins.DefaultRequest
 import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
 import io.ktor.client.request.get
 import io.ktor.client.request.parameter
+import io.ktor.client.statement.request
+import io.ktor.http.HttpStatusCode
+import io.ktor.http.URLProtocol
 import io.ktor.http.parameters
+import io.ktor.http.path
 import io.ktor.serialization.kotlinx.json.json
 import io.ktor.server.config.ApplicationConfig
 import kotlinx.serialization.Serializable
 import kotlinx.serialization.json.Json
 
+private val logger = KotlinLogging.logger {}
+
 class YoutubeClient(config: ApplicationConfig) {
     private val youtubeClient: HttpClient by lazy { prepareHttpClient() }
     private val apiKey: String? by lazy { config.propertyOrNull("youtube.apiKey")?.getString() }
 
-    suspend fun getVideoName(id: String): String {
-        val response: YoutubeVideosResponse = youtubeClient.get("/videos") {
+    suspend fun getVideoName(id: String): Result<String, YoutubeClientError> {
+        if (apiKey == null) {
+            logger.warn { "Missing Youtube API key" }
+            return Err(YoutubeClientError.ApiKeyMissing)
+        }
+
+        val response = youtubeClient.get("videos") {
             parameters {
                 parameter("id", id)
                 parameter("part", "snippet")
                 parameter("filer", "id")
             }
-        }.body()
+        }
+
+        if (response.status != HttpStatusCode.OK) {
+            logger.error { "YoutubeClient failed with status: ${response.status} \n ${response.request.url}" }
+            return Err(YoutubeClientError.UnknownError)
+        }
 
-        val video = response.items.first()
-        return video.snippet.title
+        val responseBody: YoutubeVideosResponse = response.body()
+
+        return when (val video = responseBody.items.firstOrNull()) {
+            is YoutubeVideo -> Ok(video.snippet.title)
+            else -> Err(YoutubeClientError.VideoNotFound)
+        }
     }
 
     private fun prepareHttpClient(): HttpClient {
         return HttpClient(OkHttp) {
-            install(YoutubeHttpClientPlugin) {
-                apiKey = this@YoutubeClient.apiKey
-            }
             install(ContentNegotiation) { json(Json { ignoreUnknownKeys = true }) }
+            install(DefaultRequest) {
+                url {
+                    protocol = URLProtocol.HTTPS
+                    host = "www.googleapis.com"
+                    path("/youtube/v3/")
+                    this@YoutubeClient.apiKey?.let { parameters.append("key", it) }
```

Comment:
🤔 I'm thinking aloud here.
If there is no key, we shouldn't be able to create a client. Isn't this a illegal state of application? If the key is not present, then we should not create a client?